docs/complexity_analysis.md

```diff
@@ -0,0 +1,44 @@
+# Complexity analysis of the algorithm
+
+## October 13th version
+It is a bit of a complicated task at this point to make decision on what is more impressive: the the efficiency of the algorithm or it's poor analysis.  
+I guess, it's a reason good enough to make an attempt to fix both.
+
+---
+### The algorithm explanation
+![algorithm visualization](../pictures/mawa_sort.png)
+The logic is not as sophisticated as it might first seem due to writer's unbearable stylistic choices.  
+Everything boils to these main steps:
+1. Find the middle element and check it with the searched one. We found one? The job is done.
+2. Not lucky enough? Remove the ***middle*** element. Simply eliminate it.
+3. The next step (as one could see by the source) is not really the "next" but rather a rare occasion so we'll leave it alone for a moment.
+4. Now we do another comparison of the searched element against the one to the **left** to the ***one we've just eliminated***. As the elements are in total sequential disorder, it's basically a roll of a dice. There's not much to predict of a last roll. But now we have 2 paths based on it:
+    - either we recurse into the left part of an array => (l <=> m - 1) 
+    - or we recurse into the **kinda-right-part** of an array but **not quite**, as we join the ***m - 1***th element in both scenarios => (m - 1 <=> r - 1)
+5. Just in case we've come too close to the left edge of an array, there is an additional check (see 3). Should this happen, the recursion pointers are refreshed with the arrays' edges. This saves the recursion from being indefinitely stuck.
+
+![code visualization](../pictures/mawa_sort_2.png)
+
+Let's sum up:
+- Each step 1 element is being checked for the fact of being the searched one. If it's not, it's getting eliminated.
+- We got a tottally random roll of a dice aftewards which decides whether we go left or right following.

```

Comment:
actually not? it's based whether element was smaller or not in comparison to previous eliment, algorithm might fail to go into right direction tho if there are many elements with same value in a row, needs optimization on that point possibly :|

Reply:
The average distribution of the elements in an array and an average value to search implies the following: 
- the first middle element value is totally random in comparison to the sought value;
- and basically, any other element is totally unpredictable. Any comparison is a comparison of one totally random value against the other.

It would not be true had we an already sorted array, of course. But that was not the case, was it?

Reply:
that was exactly case, this method is made for sorted arrays, and just few hacks to work on unsorted were added later.

Reply:
Had it been true, what the hell these rigged tests are then
![image](https://user-images.githubusercontent.com/93489840/139865960-309c996a-8708-4676-9001-2d4ec65df8a9.png)
aint they supposed to detect the issues with the ***unsorted*** inputs meaning, in turn, that kind of inputs had been the part of the equation?

Reply:
welp I wanted it not to crash with unsorted ones either

Reply:
then you failed as it does

Reply:
yeah, that's very true.